tests/models/client/test_llm_client.py

```diff
@@ -0,0 +1,308 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the specific language for the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Unit tests for ray_curator.stages.services.model_client module.
+"""
+
+import asyncio
+from collections.abc import Iterable
+from unittest.mock import patch
+
+import pytest
+
+from nemo_curator.models.client.llm_client import AsyncLLMClient, GenerationConfig, LLMClient
+
+
+class TestLLMClient:
+    """Test cases for the LLMClient abstract base class."""
+
+    def test_cannot_instantiate_abstract_class(self) -> None:
+        """Test that LLMClient cannot be instantiated directly."""
+        with pytest.raises(TypeError):
+            LLMClient()
+
+    def test_abstract_methods_raise_not_implemented_error(self) -> None:
+        """Test that abstract methods raise NotImplementedError when called."""
+
+        class ConcreteLLMClient(LLMClient):
+            pass
+
+        # Should not be able to instantiate without implementing abstract methods
+        with pytest.raises(TypeError):
+            ConcreteLLMClient()
+
+    def test_concrete_implementation_works(self) -> None:
+        """Test that a concrete implementation can be instantiated and used."""
+
+        class TestLLMClient(LLMClient):
+            def setup(self) -> None:
+                pass
+
+            def query_model(self, *, messages: Iterable, model: str, **kwargs: object) -> list[str]:  # noqa: ARG002
+                return ["test response"]
+
+        client = TestLLMClient()
+        client.setup()
+
+        # Test query_model
+        result = client.query_model(messages=[{"role": "user", "content": "test"}], model="test-model")
+        assert result == ["test response"]
+
+
+class TestAsyncLLMClient:
+    """Test cases for the AsyncLLMClient abstract base class."""
+
+    def test_init_with_defaults(self) -> None:
+        """Test AsyncLLMClient initialization with default parameters."""
+
+        class TestAsyncLLMClient(AsyncLLMClient):
+            def setup(self) -> None:
+                pass
+
+            async def _query_model_impl(self, *, messages: Iterable, model: str, **kwargs: object) -> list[str]:  # noqa: ARG002
+                return ["test response"]
+
+        client = TestAsyncLLMClient()
+        assert client.max_concurrent_requests == 5
+        assert client.max_retries == 3
+        assert client.base_delay == 1.0
+        assert client._semaphore is None
+        assert client._semaphore_loop is None
```

Comment:
I am not worried about checking all of these values, WDYT?

Reply:
we did this type of checking on FW side for defined recipe, but if you think its fine, I can drop them for sure 👍